File: backend/src/models/report.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Report {
    pub id: String,
    pub lat: f64,
    pub lng: f64,
    pub severity: String,
    pub photo_path: Option<String>,
    pub address: Option<String>,
    pub description: Option<String>,
    pub status: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Deserialize)]
pub struct CreateReportRequest {
    pub lat: f64,
    pub lng: f64,
    pub severity: String,
    pub photo_path: Option<String>,
    pub address: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateReportRequest {
    pub status: String,
    pub address: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ReportResponse {
    pub id: String,
    pub lat: f64,
    pub lng: f64,
    pub severity: String,
    pub photo_path: Option<String>,
    pub address: Option<String>,
    pub description: Option<String>,
    pub status: String,
    pub created_at: String,
    pub updated_at: String,
}

impl From<Report> for ReportResponse {
    fn from(r: Report) -> Self {
        Self {
            id: r.id,
            lat: r.lat,
            lng: r.lng,
            severity: r.severity,
            photo_path: r.photo_path,
            address: r.address,
            description: r.description,
            status: r.status,
            created_at: r.created_at,
            updated_at: r.updated_at,
        }
    }
}

File: android/app/src/main/java/com/jalan3d/data/Report.kt
package com.jalan3d.data

import com.jalan3d.data.api.ReportResponse

/**
 * Domain model for a road damage report.
 * Decoupled from API DTOs so the rest of the app doesn't depend on network types.
 */
data class Report(
    val id: String,
    val latitude: Double,
    val longitude: Double,
    val severity: Severity,
    val photoPath: String?,
    val address: String?,
    val description: String?,
    val status: String,
    val createdAt: String,
    val updatedAt: String
) {
    companion object {
        fun fromResponse(response: ReportResponse): Report {
            return Report(
                id = response.id,
                latitude = response.lat,
                longitude = response.lng,
                severity = Severity.fromKey(response.severity),
                photoPath = response.photoPath,
                address = response.address,
                description = response.description,
                status = response.status,
                createdAt = response.createdAt,
                updatedAt = response.updatedAt
            )
        }
    }
}
